src/types/BscType.ts

```diff
@@ -1,6 +1,45 @@
-export interface BscType {
-    isAssignableTo(targetType: BscType): boolean;
-    isConvertibleTo(targetType: BscType): boolean;
-    toString(): string;
-    toTypeString(): string;
+import type { SymbolTableProvider } from '../SymbolTable';
+import { SymbolTypeFlags } from '../SymbolTable';
+import { SymbolTable } from '../SymbolTable';
+import type { Range } from 'vscode-languageserver';
+
+export abstract class BscType {
+
+    protected symbolTable: SymbolTable;
+    protected __identifier: string;
+
+    constructor(name = '') {
+        this.__identifier = `${this.constructor.name}${name ? ': ' + name : ''}`;
+        this.symbolTable = new SymbolTable(this.__identifier);
+    }
+
+    pushMemberProvider(provider: SymbolTableProvider) {
+        this.symbolTable.pushParentProvider(provider);
+    }
+
+    popMemberProvider() {
+        this.symbolTable.popParentProvider();
+    }
+
+    addMember(name: string, range: Range, type: BscType, flags: SymbolTypeFlags) {
+        this.symbolTable.addSymbol(name, range, type, flags);
+    }
+
+    getMemberType(name: string) {
```

Comment:
Can we just call this `getMember()`? `addMember` adds a type, so `getMember` make sense for retrieving a type.

Reply:
Maybe? It's not entirely analogous, because I want to remove the boilerplate `?.[0]?.type` code...